elrond-wasm/src/storage/mappers/token_attributes_mapper.rs

```diff
@@ -0,0 +1,239 @@
+
+        !has_mapping || self.is_empty_token_attributes_value(mapping, token_nonce)
+    }
+
+    pub fn get<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+    ) -> T {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
+                .signal_error(VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE);
+        }
+
+        let value = self.get_token_attributes_value(mapping, token_nonce);
+        T::top_decode(value).unwrap()
+    }
+
+    fn build_key_token_id_counter(&self) -> Vec<u8> {
+        let mut bytes = self.main_key.as_slice().to_vec();
+        bytes.extend_from_slice(COUNTER_SUFFIX);
+        bytes
+    }
+
+    fn build_key_token_id_mapping(&self, token_id: &TokenIdentifier) -> Vec<u8> {
+        let mut bytes = self.main_key.as_slice().to_vec();
+        bytes.extend_from_slice(MAPPING_SUFFIX);
+        if let Result::Err(encode_error) = token_id.dep_encode(&mut bytes) {
+            self.api.signal_error(encode_error.message_bytes());
+        }
+        bytes
+    }
+
+    fn build_key_token_attr_value(&self, mapping: u8, token_nonce: u64) -> Vec<u8> {
+        let mut bytes = self.main_key.as_slice().to_vec();
+        if let Result::Err(encode_error) = mapping.dep_encode(&mut bytes) {
+            self.api.signal_error(encode_error.message_bytes());
+        }
+        if let Result::Err(encode_error) = token_nonce.dep_encode(&mut bytes) {
+            self.api.signal_error(encode_error.message_bytes());
+        }
+        bytes
+    }
+
+    fn get_counter_value(&self) -> u8 {
+        storage_get(
+            self.api.clone(),
+            self.build_key_token_id_counter().as_slice(),
+        )
+    }
+
+    fn set_counter_value(&self, value: u8) {
+        storage_set(
+            self.api.clone(),
+            self.build_key_token_id_counter().as_slice(),
+            &value,
+        );
+    }
+
+    fn get_mapping_value(&self, token_id: &TokenIdentifier) -> u8 {
+        storage_get(
+            self.api.clone(),
+            self.build_key_token_id_mapping(token_id).as_slice(),
+        )
+    }
+
+    fn set_mapping_value(&self, token_id: &TokenIdentifier, value: u8) {
+        storage_set(
+            self.api.clone(),
+            self.build_key_token_id_mapping(token_id).as_slice(),
+            &value,
+        );
+    }
+
+    fn is_empty_mapping_value(&self, token_id: &TokenIdentifier) -> bool {
+        self.api
+            .storage_load_len(self.build_key_token_id_mapping(token_id).as_slice())
+            == 0
+    }
+
+    fn get_token_attributes_value(&self, mapping: u8, token_nonce: u64) -> Vec<u8> {
```

Comment:
Make this generic.
get_token_attributes_value<T: TopEncode + TopDecode>

Reply:
The idea was to have a storage for all types of custom Attrs struct